src/python/pants/engine/native.py

```diff
@@ -29,251 +30,6 @@
-void tasks_singleton_add(Tasks*, Handle, TypeConstraint);
-void tasks_destroy(Tasks*);
-
-Scheduler* scheduler_create(Tasks*,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            Function,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeConstraint,
-                            TypeId,
-                            TypeId,
-                            Buffer,
-                            Buffer,
-                            Buffer,
-                            BufferBuffer,
-                            TypeIdBuffer,
-                            Buffer,
-                            Buffer,
-                            Buffer,
-                            Buffer,
-                            Buffer,
-                            Buffer,
-                            uint64_t,
-                            uint64_t,
-                            uint64_t,
-                            uint64_t,
-                            _Bool);
-PyResult scheduler_fork_context(Scheduler*, Function);
-Handle scheduler_metrics(Scheduler*, Session*);
-RawNodes* scheduler_execute(Scheduler*, Session*, ExecutionRequest*);
-void scheduler_destroy(Scheduler*);
-
-Session* session_create(Scheduler*);
-void session_destroy(Session*);
-
-ExecutionRequest* execution_request_create(_Bool, uint64_t);
-void execution_request_destroy(ExecutionRequest*);
-
-uint64_t graph_len(Scheduler*);
-uint64_t graph_invalidate(Scheduler*, BufferBuffer);
-uint64_t graph_invalidate_all_paths(Scheduler*);
-PyResult graph_visualize(Scheduler*, Session*, char*);
-void graph_trace(Scheduler*, ExecutionRequest*, char*);
-
-PyResult execution_add_root_select(Scheduler*, ExecutionRequest*, HandleBuffer, TypeConstraint);
-
-PyResult capture_snapshots(Scheduler*, Handle);
-
-PyResult merge_directories(Scheduler*, Handle);
-
-PyResult materialize_directories(Scheduler*, Handle);
-
-PyResult validator_run(Scheduler*);
-
-void rule_graph_visualize(Scheduler*, TypeIdBuffer, char*);
-void rule_subgraph_visualize(Scheduler*, TypeId, TypeConstraint, char*);
-
-void nodes_destroy(RawNodes*);
-
-void set_panic_handler(void);
-
-void lease_files_in_graph(Scheduler*);
-
-void garbage_collect_store(Scheduler*);
```

Comment:
Q: Using `garbage_collect_store` as an example, is it needed to have its corresponding changes, or is it a feature of the new lib as well?

Reply:
So the usage of the `cbindgen` crate will produce this file `src/rust/engine/src/cffi/scheduler.h` when pants is bootstrapped from this repo -- in this case, that generated header includes the line:
```C
void garbage_collect_store(Scheduler *scheduler_ptr);
```

So the idea is that we can hopefully rely on `cbindgen` to do this for us, or better. It is good that the result (the generated `scheduler.h` header) is compatible with `cffi` -- I expect this to continue to be true, but any new dependency has risks.

Reply:
Well, actually, the generated header already isn't immediately compatible with `cffi`, and we are already doing some regex substitution hacks to get around this -- and any further updates to `cffi` would be to *support* more of normal header files such as `#include` lines, I think, not less.